crates/oxc_macros/src/declare_all_lint_rules.rs

```diff
@@ -0,0 +1,125 @@
+use convert_case::{Case, Casing};
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::parse::{Parse, ParseStream};
+use syn::Result;
+
+pub struct LintRuleMeta {
+    name: syn::Ident,
+    path: syn::Path,
+}
+
+impl LintRuleMeta {
+    pub fn mod_stmt(&self) -> TokenStream {
+        let mut segments = self.path.segments.iter().rev().peekable();
+        let first = &segments.next().unwrap().ident;
+        let mut stmts = quote! {mod #first;};
+        if segments.peek().is_some() {
+            stmts = quote! {pub #stmts};
+        }
+
+        while let Some(segment) = segments.next() {
+            let ident = &segment.ident;
+
+            stmts = quote! {
+                mod #ident { #stmts }
+            };
+
+            if segments.peek().is_some() {
+                stmts = quote! {
+                    pub #stmts
+                };
+            }
+        }
+
+        stmts
+    }
+
+    pub fn use_stmt(&self) -> TokenStream {
+        let mut path = self.path.clone();
+        path.segments.push(self.name.clone().into());
+
+        quote! {
+            pub use #path;
+        }
+    }
+}
+
+impl Parse for LintRuleMeta {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let path = input.parse::<syn::Path>()?;
+        let name = syn::parse_str(
+            &path.segments.iter().last().unwrap().ident.to_string().to_case(Case::Pascal),
+        )
+        .unwrap();
+        Ok(Self { path, name })
+    }
+}
+
+pub struct AllLintRulesMeta {
+    rules: Vec<LintRuleMeta>,
+}
+
+impl Parse for AllLintRulesMeta {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let rules = input
+            .parse_terminated::<LintRuleMeta, syn::Token![,]>(LintRuleMeta::parse)?
+            .into_iter()
+            .collect();
+
+        Ok(Self { rules })
+    }
+}
+
+pub fn declare_all_lint_rules(metadata: AllLintRulesMeta) -> TokenStream {
+    let AllLintRulesMeta { rules } = metadata;
+
+    let mod_stmts = rules.iter().map(|rule| rule.mod_stmt());
+    let use_stmts = rules.iter().map(|rule| rule.use_stmt());
+    let names = rules.iter().map(|rule| &rule.name);
+    let names2 = names.clone();
+    let names3 = names.clone();
+    let names4 = names.clone();
+    let names5 = names.clone();

```

Comment:
This is pretty terrible, but I honestly wasn't sure how to do better 😓

Reply:
Should this be called `struct_names`? And I don't think you need to clone it, `syn` should be able to use it repetitively (I think).

Reply:
Oh! you just need to collect `names` into a Vec!

Reply:
oh wow I'm dumb